### ISSUE TITLE
about UV Total Variation Loss

### ISSUE THREAD

**Issue by lizhaohu** [tobias-kirschstein/gghead#4]
Thank you for sharing the code. Could you please point me to where the UV Total Variation Loss mentioned in the paper is implemented in the code?

**Comment by tobias-kirschstein** [comment]
Hi,
thank you for your interest in our work.
The implementation of the UV Total Variation Loss will soon come together with the training code release.
In the meanwhile, this is the relevant part of the loss calculation that should get you going:

```python
uv_renderings = model.get_uv_rendering(gen_c, gen_img)
mask = 1 - (uv_renderings[:, [2]] + 1)/2

uv_renderings_no_blend = ((uv_renderings - (1 - mask)) / mask)  # Undo effect of alpha blending. Background pixels will be inf
background_mask = mask == 0

mask_y = (background_mask[:, :, 1:] | background_mask[:, :, :-1]).repeat(1, 3, 1, 1)
mask_x = (background_mask[:, :, :, 1:] | background_mask[:, :, :, :-1]).repeat(1, 3, 1, 1)

uv_difference_y = uv_renderings_no_blend[:, :, 1:] - uv_renderings_no_blend[:, :, :-1]
uv_difference_x = uv_renderings_no_blend[:, :, :, 1:] - uv_renderings_no_blend[:, :, :, :-1]

reg_tv_uv_rendering_y = uv_difference_y[~mask_y].abs().mean()
reg_tv_uv_rendering_x = uv_difference_x[~mask_x].abs().mean()
reg_tv_uv_rendering = reg_tv_uv_rendering_x + reg_tv_uv_rendering_y

loss_Gmain = loss_Gmain + lambda_tv_uv_rendering * reg_tv_uv_rendering
```

The `get_uv_rendering()` method is part of the `GGHeadModel` class. The rest is a fairly standard implementation of a total variation computation. We used `lambda_tv_uv_rendering = 100`.
Hope, this helps.

**Comment by lizhaohu** [comment]
> Hi, thank you for your interest in our work. The implementation of the UV Total Variation Loss will soon come together with the training code release. In the meanwhile, this is the relevant part of the loss calculation that should get you going:
> 
> ```python
> uv_renderings = model.get_uv_rendering(gen_c, gen_img)
> mask = 1 - (uv_renderings[:, [2]] + 1)/2
> 
> uv_renderings_no_blend = ((uv_renderings - (1 - mask)) / mask)  # Undo effect of alpha blending. Background pixels will be inf
> background_mask = mask == 0
> 
> mask_y = (background_mask[:, :, 1:] | background_mask[:, :, :-1]).repeat(1, 3, 1, 1)
> mask_x = (background_mask[:, :, :, 1:] | background_mask[:, :, :, :-1]).repeat(1, 3, 1, 1)
> 
> uv_difference_y = uv_renderings_no_blend[:, :, 1:] - uv_renderings_no_blend[:, :, :-1]
> uv_difference_x = uv_renderings_no_blend[:, :, :, 1:] - uv_renderings_no_blend[:, :, :, :-1]
> 
> reg_tv_uv_rendering_y = uv_difference_y[~mask_y].abs().mean()
> reg_tv_uv_rendering_x = uv_difference_x[~mask_x].abs().mean()
> reg_tv_uv_rendering = reg_tv_uv_rendering_x + reg_tv_uv_rendering_y
> 
> loss_Gmain = loss_Gmain + lambda_tv_uv_rendering * reg_tv_uv_rendering
> ```
> 
> The `get_uv_rendering()` method is part of the `GGHeadModel` class. The rest is a fairly standard implementation of a total variation computation. We used `lambda_tv_uv_rendering = 100`. Hope, this helps.


Thank you very much. We look forward to your release of the training code
